proposals/0621-linear-constraints.rst

```diff
@@ -0,0 +1,1108 @@
+  dup2LinearityToken :: LinearityToken %1 -> (# LinearityToken, LinearityToken #)
+
+  data DupableClassModifier = Dupable
+
+Class declaration can be annotated with the ``%Dupable``
+
+::
+
+   %Dupable class <ctx> => <head> where
+    <methods>
+
+Classes annotated with ``%Dupable`` must:
+
+- Have a single method
+- The method must be of type ``LinearityToken``
+
+Such a dupable type class can be used multiple times even if they are
+linear. *E.g.*::
+
+  class Foo where
+    foo' :: LinearityToken
+
+  foo :: Foo %1 => Int -> Int
+  foo = case foo' of { (# #) -> \x -> x }
+
+  dupes :: Foo %1 => (Int -> Int, Int -> Int)
+  dupes = (foo, foo)
+
+But, crucially, not passed to an unrestricted function::
+
+  rejected :: Foo %1 => Ur (Int -> Int)
+  rejected = Ur foo
+
+There are a lot of alternatives for the design of this feature, so see
+the *Alternatives* section for more thoughts.
+
+Examples
+--------
+
+Threading
+^^^^^^^^^
+
+We refer back now to the first example from the motivation section,
+which showed how writing a function which reads the first two elements
+of an array became a tedious exercise of threading our linear resource
+through the function. Using linear constraints, however, such a
+function can be written as:
+
+::
+
+   read2AndDiscard ::  (Read n, Write n) %1 => MArray a n -> (Ur a, Ur a)
+   read2AndDiscard arr =
+        let !(Box x)  = read arr 0
+            !(Box y)  = read arr 1
+            !()       = free arr
+        in (x, y)
+
+The main way in which this differs from our previous function is that
+our array is no longer a linear resource - it is
+*unrestricted*. However, we maintain the guarantee that it is used in
+a way which does not violate linearity through the ``Read n`` and
+``Write n`` linear constraints. Here, ``n`` is a type-level tag used
+to identify the array. Accordingly, our type constructor for
+``MArray`` is parameterised by ``n``.
+
+The type signatures for  ``read``, ``free``, and ``Box`` are:
+
+::
+
+   read  :: Read n %1 => MArray a n -> Int -> Read n /\ Ur a
+
+   free :: (Read n, Write n) %1 => MArray a n -> ()
+
+   data c /\ a where
+     Box :: c %1 => a -> c /\ a
+
+i.e. ``read`` is a function which consumes a linear ``Read n``
+constraint, allowing us to read from the specified array index. It also
+returns a new ``Read n`` constraint, allowing us to subsequently read
+from the array again. Likewise, ``free`` consumes both a ``Read n``
+and a ``Write n`` constraint and introduces none, ensuring that we
+cannot read or write after freeing.
+
+Thus we eliminate the need to manually thread the ownership of the
+array through the function, whilst maintaining the guarantees of
+unique ownership via the linear constraints.
+
+For a more in-depth example along these lines, refer to section 4 of
+the paper_.
+
+The Linearly constraint
+^^^^^^^^^^^^^^^^^^^^^^^
+
+We can create a class, the paper_ calls it ``Linearly`` with the
+following API:
+
+::
+
+  %Dupable class Linearly
```

Comment:
If, as mentioned below, all `%Dupable` classes are isomorphic, wouldn't it make sense to also include that class in `GHC.Constraint.Linear`?